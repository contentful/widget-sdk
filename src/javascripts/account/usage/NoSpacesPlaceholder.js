import React from 'react';

export default () => (
  <header className="usage-page__no-spaces-placeholder">
    Usage will be available after you created your first space.
  </header>
);
