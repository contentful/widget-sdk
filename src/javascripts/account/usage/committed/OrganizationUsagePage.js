import React from 'react';
import PropTypes from 'prop-types';
import { map, sum, get } from 'lodash';
import * as Config from 'Config';
import { getCurrentStateName } from 'states/Navigator';

import {
  organizationResourceUsagePropType,
  organizationUsagePropType,
  arrayPropType,
  periodPropType
} from './propTypes';
import OrganizationUsageInfo from './OrganizationUsageInfo';
import AssetBandwidthSection from './AssetBandwidthSection';
import OrganizationUsageChart from './charts/OrganizationUsageChart';
import ApiUsageInfo from './ApiUsageInfo';
import ApiUsageChart from './charts/ApiUsageChart';
import * as Intercom from 'services/intercom';
import * as Analytics from 'analytics/Analytics';
import { seriesAppearance } from './charts/constants';

const apiUsagePropType = arrayPropType(organizationResourceUsagePropType);

export default class OrganizationUsagePage extends React.Component {
  static propTypes = {
    spaceNames: PropTypes.objectOf(PropTypes.string).isRequired,
    isPoC: PropTypes.objectOf(PropTypes.bool).isRequired,
    periodicUsage: PropTypes.shape({
      org: organizationUsagePropType,
      apis: PropTypes.shape({
        cma: apiUsagePropType,
        cda: apiUsagePropType,
        cpa: apiUsagePropType,
        gql: apiUsagePropType
      })
    }).isRequired,
    period: periodPropType,
    apiRequestIncludedLimit: PropTypes.number.isRequired,
    assetBandwidthData: PropTypes.shape({
      usage: PropTypes.number,
      unitOfMeasure: PropTypes.string,
      limits: PropTypes.shape({
        included: PropTypes.number
      })
    }),
    isLoading: PropTypes.bool.isRequired
  };

  onClickSupport = () => {
    Analytics.track('element:click', {
      elementId: 'contact_sales_usage',
      groupId: 'contact_sales',
      fromState: getCurrentStateName()
    });
    Intercom.isEnabled() ? Intercom.open() : window.open(Config.supportUrl);
  };

  render() {
    const {
      periodicUsage: {
        org: { usage: orgUsage },
        apis
      },
      apiRequestIncludedLimit,
      assetBandwidthData,
      spaceNames,
      period,
      isLoading,
      isPoC
    } = this.props;

    const totalUsage = sum(orgUsage);

    return (
      <div className="usage-page">
        <div className="usage-page__section">
          <OrganizationUsageInfo totalUsage={totalUsage} includedLimit={apiRequestIncludedLimit} />
          <OrganizationUsageChart
            usage={orgUsage}
            includedLimit={apiRequestIncludedLimit}
            period={period}
            isLoading={isLoading}
          />
        </div>
        {map(apis, (usage, api) => (
          <div key={api} className="usage-page__section">
            <ApiUsageInfo
              totalUsage={totalUsage}
              api={api}
              spaceNames={spaceNames}
              isPoC={isPoC}
              usage={usage.items}
              colors={seriesAppearance.map(item => item.color)}
            />
            <ApiUsageChart
              usage={usage.items}
              period={period}
              spaceNames={spaceNames}
              isLoading={isLoading}
            />
          </div>
        ))}
        <AssetBandwidthSection
          limit={get(assetBandwidthData, ['limits', 'included'])}
          usage={get(assetBandwidthData, ['usage'])}
          uom={get(assetBandwidthData, ['unitOfMeasure'])}
        />
      </div>
    );
  }
}
