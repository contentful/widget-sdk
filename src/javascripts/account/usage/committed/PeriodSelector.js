import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';

import { periodPropType } from './propTypes.es6';

const formatDate = date => moment(date).format('DD MMM');

export default class PeriodSelector extends React.Component {
  static propTypes = {
    periods: PropTypes.arrayOf(periodPropType).isRequired,
    selectedPeriodIndex: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
  };

  render() {
    const { periods, onChange, selectedPeriodIndex } = this.props;
    return (
      <span className="usage__period-selector">
        <label>Usage period</label>
        <Select
          defaultValue={selectedPeriodIndex}
          onChange={onChange}
          width="auto"
          name="period-selector"
          id="period-selector">
          {periods.map(({ startDate, endDate }, index) => {
            const isCurrentPeriod = endDate === null;
            const start = moment(startDate);
            const end = isCurrentPeriod
              ? moment(start)
                  .add(1, 'month')
                  .subtract(1, 'day')
              : moment(endDate);

            const value = `${index}`;

            return (
              <Option key={index} value={value}>{`${formatDate(start)} â€“ ${formatDate(end)} ${cond([
                [constant(isCurrentPeriod), constant('(current)')],
                [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
                [constant(stubTrue), constant(end.year())]
              ])()}`}</Option>
            );
          })}
        </Select>
      </span>
    );
  }
}
