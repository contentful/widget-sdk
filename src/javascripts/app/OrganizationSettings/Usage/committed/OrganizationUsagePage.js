import React from 'react';
import MainTabs from './tabs/MainTabs';

import PropTypes from 'prop-types';
import { periodicUsagePropType, periodPropType } from './propTypes';

import periodToDates from './charts/periodToDates';

const OrganizationUsagePage = ({
  spaceNames,
  period,
  periodicUsage,
  apiRequestIncludedLimit,
  assetBandwidthData,
  onTabSelect,
  isPoC
}) => {
  return (
    <MainTabs
      assetBandwidthData={assetBandwidthData}
      period={periodToDates(period)}
      periodicUsage={periodicUsage}
      apiRequestIncludedLimit={apiRequestIncludedLimit}
      spaceNames={spaceNames}
      isPoC={isPoC}
      onTabSelect={onTabSelect}
    />
  );
};

OrganizationUsagePage.propTypes = {
  spaceNames: PropTypes.objectOf(PropTypes.string).isRequired,
  periodicUsage: periodicUsagePropType.isRequired,
  period: periodPropType,
  apiRequestIncludedLimit: PropTypes.number.isRequired,
  assetBandwidthData: PropTypes.shape({
    usage: PropTypes.number,
    unitOfMeasure: PropTypes.string,
    limits: PropTypes.shape({
      included: PropTypes.number
    })
  }),
  isPoC: PropTypes.objectOf(PropTypes.bool).isRequired,
  onTabSelect: PropTypes.func
};

export default OrganizationUsagePage;
