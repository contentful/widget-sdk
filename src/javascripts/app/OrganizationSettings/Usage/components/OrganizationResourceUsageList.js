import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import { keyBy, property } from 'lodash';
import { ResourceUsage } from 'app/SpaceSettings/Usage/ResourceUsage';
import { Paragraph, Heading } from '@contentful/forma-36-react-components';

export default class OrganizationResourceUsageList extends React.Component {
  static propTypes = {
    resources: PropTypes.arrayOf(PropTypes.object)
  };

  render() {
    const { resources } = this.props;
    if (!resources.length) return null;

    const formatDate = date => moment(date).format('DD MMMM');

    const byId = keyBy(resources, property('sys.id'));
    const { start, end } = byId['api_request'].period;
    const nowNormalized = moment.utc(new Date().toDateString(), 'ddd MMM DD YYYY');
    const daysLeft = moment(end).diff(nowNormalized, 'days');
    const startDate = formatDate(start);
    const endDate = formatDate(end);

    return (
      <div className="resource-list">
        <ResourceUsage resource={byId['organization_membership']} />
        <div className="resource-list__title">
          <Heading element="h3" className="section-title">
            Current billing period
          </Heading>
          <Paragraph>
            {startDate} â€“ {endDate} ({daysLeft} days remaining)
          </Paragraph>
        </div>
        <ResourceUsage resource={byId['api_request']} abbreviateLimit={true} />
        <ResourceUsage resource={byId['asset_bandwidth']} />
      </div>
    );
  }
}
