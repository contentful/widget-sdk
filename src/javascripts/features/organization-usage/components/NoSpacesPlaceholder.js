import React from 'react';
import EmptyStateContainer, {
  defaultSVGStyle,
} from 'components/EmptyStateContainer/EmptyStateContainer';
import EmptyStateAdminIllustration from 'svg/folder-illustration.svg';
import { Heading, Paragraph, Button, TextLink } from '@contentful/forma-36-react-components';

import * as CreateSpace from 'services/CreateSpace';
import { useUsageState } from '../hooks/usageContext';

export const NoSpacesPlaceholder = () => {
  const { orgId } = useUsageState();
  return (
    <EmptyStateContainer data-test-id="usage-page__no-spaces-placeholder">
      <EmptyStateAdminIllustration className={defaultSVGStyle} />
      <Heading>Looking for your usage data?</Heading>
      <Paragraph>
        Looks like you need to create a space first. Whatâ€™s a space? Good question! A space is an
        area to manage and store content for a specific project.
      </Paragraph>
      <Paragraph>
        You can learn more about spaces in the{' '}
        <TextLink
          href="https://www.contentful.com/developers/docs/concepts/domain-model/"
          target="_blank"
          rel="noopener noreferrer">
          domain model documentation
        </TextLink>
        .
      </Paragraph>
      <Button onClick={() => CreateSpace.showDialog(orgId)}>Add a space</Button>
    </EmptyStateContainer>
  );
};
