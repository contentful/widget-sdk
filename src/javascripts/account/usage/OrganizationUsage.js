import React from 'react';
import PropTypes from 'prop-types';
import { mapValues, flow, keyBy, get, eq, isNumber, pick } from 'lodash/fp';

import { Spinner, Workbench } from '@contentful/forma-36-react-components';
import ReloadNotification from 'app/common/ReloadNotification';

import OrganizationResourceUsageList from './non_committed/OrganizationResourceUsageList';
import OrganizationUsagePage from './committed/OrganizationUsagePage';
import PeriodSelector from './committed/PeriodSelector';
import NoSpacesPlaceholder from './NoSpacesPlaceholder';
import * as Analytics from 'analytics/Analytics';
import * as UsageService from '././UsageService';

import * as TokenStore from 'services/TokenStore';
import * as EndpointFactory from 'data/EndpointFactory';
import * as OrganizationMembershipRepository from 'access_control/OrganizationMembershipRepository';
import * as PricingDataProvider from 'account/pricing/PricingDataProvider';
import createResourceService from 'services/ResourceService';
import * as OrganizationRoles from 'services/OrganizationRoles';
import Icon from 'ui/Components/Icon';
import ErrorState from 'app/common/ErrorState';

export class WorkbenchContent extends React.Component {
  static propTypes = {
    committed: PropTypes.bool,
    hasSpaces: PropTypes.bool,
    selectedPeriodIndex: PropTypes.number,
    spaceNames: PropTypes.objectOf(PropTypes.string),
    isPoC: PropTypes.objectOf(PropTypes.bool),
    periodicUsage: PropTypes.object,
    apiRequestIncludedLimit: PropTypes.number,
    assetBandwidthData: PropTypes.shape({
      usage: PropTypes.number,
      unitOfMeasure: PropTypes.string,
      limits: PropTypes.shape({
        included: PropTypes.number
      })
    }),
    isLoading: PropTypes.bool,
    error: PropTypes.string,
    periods: PropTypes.arrayOf(PropTypes.object),
    resources: PropTypes.arrayOf(PropTypes.object)
  };

  render() {
    const {
      committed,
      hasSpaces,
      selectedPeriodIndex,
      spaceNames,
      isPoC,
      periodicUsage,
      apiRequestIncludedLimit,
      assetBandwidthData,
      isLoading,
      error,
      periods,
      resources
    } = this.props;

    if (error) {
      return <ErrorState />;
    }

    if (committed) {
      if (!hasSpaces) {
        return <NoSpacesPlaceholder />;
      }
      if (typeof selectedPeriodIndex !== 'undefined') {
        return (
          <OrganizationUsagePage
            {...{
              period: periods[selectedPeriodIndex],
              spaceNames,
              isPoC,
              periodicUsage,
              apiRequestIncludedLimit,
              assetBandwidthData,
              isLoading
            }}
          />
        );
      }
    } else {
      if (typeof resources !== 'undefined') {
        return <OrganizationResourceUsageList resources={resources} />;
      }
    }
    return <div />;
  }
}

export class WorkbenchActions extends React.Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    error: PropTypes.string,
    hasSpaces: PropTypes.bool,
    committed: PropTypes.bool,
    periods: PropTypes.array,
    selectedPeriodIndex: PropTypes.number,
    setPeriodIndex: PropTypes.func
  };

  render() {
    const {
      isLoading,
      error,
      hasSpaces,
      committed,
      periods,
      selectedPeriodIndex,
      setPeriodIndex
    } = this.props;

    if (error) {
      return null;
    }

    if (isLoading) {
      return <Spinner />;
    }

    if (hasSpaces && committed && periods) {
      return (
        <PeriodSelector
          periods={periods}
          selectedPeriodIndex={selectedPeriodIndex}
          onChange={setPeriodIndex}
        />
      );
    }

    return null;
  }
}

export class OrganizationUsage extends React.Component {
  static propTypes = {
    orgId: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);

    this.state = { isLoading: true, error: null };

    this.endpoint = EndpointFactory.createOrganizationEndpoint(props.orgId);
    this.setPeriodIndex = this.setPeriodIndex.bind(this);
  }

  async componentDidMount() {
    try {
      await this.checkPermissions();
      await this.fetchOrgData();
    } catch (ex) {
      this.setState({ isLoading: false, error: ex.message });
    }
  }

  async checkPermissions() {
    const { orgId } = this.props;
    const organization = await TokenStore.getOrganization(orgId);

    if (!OrganizationRoles.isOwnerOrAdmin(organization)) {
      throw new Error('No permission');
    }
  }

  async fetchOrgData() {
    const { orgId } = this.props;

    try {
      const service = createResourceService(orgId, 'organization');
      const basePlan = await PricingDataProvider.getBasePlan(this.endpoint);
      const committed = PricingDataProvider.isEnterprisePlan(basePlan);

      if (committed) {
        const [
          spaces,
          plans,
          periods,
          {
            limits: { included: apiRequestIncludedLimit }
          }
        ] = await Promise.all([
          OrganizationMembershipRepository.getAllSpaces(this.endpoint),
          PricingDataProvider.getPlansWithSpaces(this.endpoint),
          UsageService.getPeriods(this.endpoint),
          service.get('api_request')
        ]);
        const spaceNames = flow(
          keyBy('sys.id'),
          mapValues('name')
        )(spaces);

        const isPoC = flow(
          keyBy('space.sys.id'),
          mapValues(
            flow(
              get('name'),
              eq('Proof of concept')
            )
          )
        )(plans.items);

        this.setState({
          spaceNames,
          isPoC,
          periods: periods.items,
          apiRequestIncludedLimit,
          hasSpaces: spaces.length !== 0
        });

        await this.loadPeriodData(0);
      } else {
        this.setState({ resources: await service.getAll(), isLoading: false });
      }

      this.setState({ committed });
    } catch (e) {
      // Show the forbidden screen on 404 and 403
      if ([404, 403].includes(e.status)) {
        throw e;
      }

      ReloadNotification.trigger();
    }
  }

  loadPeriodData = async newIndex => {
    const { orgId } = this.props;
    const { periods } = this.state;

    const service = createResourceService(orgId, 'organization');
    const newPeriod = periods[newIndex];

    if (isNumber(this.state.selectedPeriodIndex)) {
      const oldPeriod = periods[this.state.selectedPeriodIndex];
      Analytics.track('usage:period_selected', {
        oldPeriod: pick(['startDate', 'endDate'], oldPeriod),
        newPeriod: pick(['startDate', 'endDate'], newPeriod)
      });
    }

    try {
      const promises = [
        UsageService.getOrgUsage(this.endpoint, {
          startDate: newPeriod.startDate,
          endDate: newPeriod.endDate,
          periodId: newPeriod.sys.id
        }),
        ...['cma', 'cda', 'cpa', 'gql'].map(apiType =>
          UsageService.getApiUsage(this.endpoint, {
            apiType,
            startDate: newPeriod.startDate,
            endDate: newPeriod.endDate,
            periodId: newPeriod.sys.id
          })
        )
      ];

      if (newIndex === 0) {
        promises.push(service.get('asset_bandwidth'));
      } else {
        // If the current usage period is not the first (current/most recent), we return null
        // for the asset bandwidth information.
        //
        // We don't have historical (or day-to-day) AB usage and can only show data for the user's
        // current usage period.
        promises.push(Promise.resolve(null));
      }

      const [org, cma, cda, cpa, gql, assetBandwidthData] = await Promise.all(promises);

      this.setState(
        UsageService.mapResponseToState({
          org,
          cma,
          cda,
          cpa,
          gql,
          assetBandwidthData,
          newIndex
        })
      );
    } catch (e) {
      ReloadNotification.trigger();
    }
  };

  async setPeriodIndex(e) {
    this.setState({ isLoading: true });
    await this.loadPeriodData(parseInt(e.target.value));
  }

  render() {
    const {
      spaceNames,
      isPoC,
      selectedPeriodIndex,
      isLoading,
      error,
      periods,
      periodicUsage,
      apiRequestIncludedLimit,
      assetBandwidthData,
      committed,
      resources,
      hasSpaces
    } = this.state;
    return (
      <Workbench testId="organization.usage">
        <Workbench.Header
          title="Usage"
          icon={<Icon name="page-usage" scale="0.75" />}
          actions={
            <WorkbenchActions
              {...{
                isLoading,
                error,
                hasSpaces,
                committed,
                periods,
                selectedPeriodIndex,
                setPeriodIndex: this.setPeriodIndex
              }}
            />
          }></Workbench.Header>
        <Workbench.Content>
          <WorkbenchContent
            {...{
              committed,
              hasSpaces,
              selectedPeriodIndex,
              spaceNames,
              isPoC,
              periodicUsage,
              apiRequestIncludedLimit,
              assetBandwidthData,
              isLoading,
              error,
              periods,
              resources
            }}
          />
        </Workbench.Content>
      </Workbench>
    );
  }
}

export default OrganizationUsage;
