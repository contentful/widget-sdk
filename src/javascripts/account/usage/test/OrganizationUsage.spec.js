import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { isEqual } from 'lodash';

import { Spinner } from '@contentful/ui-component-library';

import { OrganizationUsage, WorkbenchContent, WorkbenchActions } from '../OrganizationUsage.es6';
import PeriodSelector from '../committed/PeriodSelector.es6';
import NoSpacesPlaceholder from '../NoSpacesPlaceholder.es6';
import OrganizationUsagePage from '../committed/OrganizationUsagePage.es6';
import OrganizationResourceUsageList from 'account/usage/non_committed/OrganizationResourceUsageList.es6';

const DATE_FORMAT = 'YYYY-MM-DD';

let defaultProps = null;
let testOrg = null;
let endpoint = null;
let resourceService = null;
const shallowRenderComponent = async props => {
  const wrapper = shallow(<OrganizationUsage {...props} />);
  // Need to wait for internal async logic to finish
  // This means `componentDidMount` is called twice,
  //  which could potentially cause problems if it's not
  //  idempotent.
  await wrapper.instance().componentDidMount();
  return wrapper;
};

describe('OrganizationUsage', () => {
  beforeAll(() => {
    // set fixed date for stable snapshots
    // moment('2017-12-01').unix() = 1512082800
    jest.spyOn(Date, 'now').mockImplementation(() => 1512082800);
  });

  afterAll(() => {
    Date.now.mockRestore();
  });

  beforeEach(() => {
    testOrg = {};
    const startDate = moment().subtract(12, 'days');

    endpoint = jest.fn(({ method, path }) => {
      if (method === 'GET' && isEqual(path, ['usage_periods'])) {
        return {
          items: [
            {
              startDate: startDate.format(DATE_FORMAT),
              endDate: null,
              sys: { type: 'UsagePeriod', id: '0' }
            },
            {
              startDate: moment(startDate)
                .subtract(1, 'day')
                .subtract(1, 'month')
                .format(DATE_FORMAT),
              endDate: moment(startDate)
                .subtract(1, 'day')
                .format(DATE_FORMAT),
              sys: { type: 'UsagePeriod', id: '1' }
            }
          ]
        };
      }
    });

    const isEnterprisePlan = jest.fn(() => true);
    const isOwnerOrAdmin = jest.fn(() => true);
    const getPlansWithSpaces = jest.fn(() => ({
      items: [
        { name: 'Test plan', space: { sys: { id: 'space1' } } },
        { name: 'Proof of concept', space: { sys: { id: 'space2' } } }
      ]
    }));

    resourceService = {
      get: jest.fn(resource => {
        switch (resource) {
          case 'api_request':
            return { limits: { included: 1000000 } };
          case 'asset_bandwidth':
            return {
              usage: 200,
              unitOfMeasure: 'MB',
              limits: { included: 2000 }
            };
        }
      }),
      getAll: jest.fn()
    };
    const ResourceService = { default: () => resourceService };

    const getAllSpaces = jest.fn(() => [
      { name: 'Test1', sys: { id: 'test1' } },
      { name: 'Test2', sys: { id: 'test2' } }
    ]);

    const getCurrentVariation = jest.fn(fs => fs === 'feature-bizvel-09-2018-usage');

    const getOrganization = jest.fn(() => testOrg);

    defaultProps = {
      orgId: '23423',
      onReady: jest.fn(),
      onForbidden: jest.fn(),
      $services: {
        OrganizationRoles: { isOwnerOrAdmin },
        PricingDataProvider: {
          isEnterprisePlan,
          getBasePlan: jest.fn(),
          getPlansWithSpaces
        },
        ResourceService,
        ReloadNotification: { trigger: jest.fn() },
        OrganizationMembershipRepository: { getAllSpaces },
        EndpointFactory: { createOrganizationEndpoint: () => endpoint },
        Analytics: { track: jest.fn() },
        LaunchDarkly: { getCurrentVariation },
        TokenStore: { getOrganization }
      }
    };
  });

  it('should render page without errors', async () => {
    const wrapper = await shallowRenderComponent(defaultProps);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call `onReady`', async () => {
    await shallowRenderComponent(defaultProps);
    expect(defaultProps.onReady).toHaveBeenCalled();
  });

  it('should request data', async () => {
    await shallowRenderComponent(defaultProps);

    expect(endpoint).toHaveBeenCalledWith(
      {
        method: 'GET',
        path: ['usage_periods']
      },
      { 'x-contentful-enable-alpha-feature': 'usage-insights' }
    );
    expect(endpoint).toHaveBeenCalledWith(
      {
        method: 'GET',
        path: ['usages', 'organization'],
        query: {
          'filters[metric]': 'allApis',
          'filters[usagePeriod]': '0'
        }
      },
      { 'x-contentful-enable-alpha-feature': 'usage-insights' }
    );
    expect(endpoint).toHaveBeenCalledWith(
      {
        method: 'GET',
        path: ['usages', 'space'],
        query: {
          'filters[metric]': 'cpa',
          'filters[usagePeriod]': '0',
          'orderBy[metricUsage]': 'desc',
          limit: 3
        }
      },
      { 'x-contentful-enable-alpha-feature': 'usage-insights' }
    );
    expect(endpoint).toHaveBeenCalledWith(
      {
        method: 'GET',
        path: ['usages', 'space'],
        query: {
          'filters[metric]': 'cda',
          'filters[usagePeriod]': '0',
          'orderBy[metricUsage]': 'desc',
          limit: 3
        }
      },
      { 'x-contentful-enable-alpha-feature': 'usage-insights' }
    );
    expect(endpoint).toHaveBeenCalledWith(
      {
        method: 'GET',
        path: ['usages', 'space'],
        query: {
          'filters[metric]': 'cma',
          'filters[usagePeriod]': '0',
          'orderBy[metricUsage]': 'desc',
          limit: 3
        }
      },
      { 'x-contentful-enable-alpha-feature': 'usage-insights' }
    );
  });

  describe('user is not owner or admin', () => {
    beforeEach(() => {
      defaultProps.$services.OrganizationRoles.isOwnerOrAdmin.mockReturnValue(false);
    });

    it('should call `onForbidden`', async () => {
      await shallowRenderComponent(defaultProps);

      expect(defaultProps.$services.TokenStore.getOrganization).toHaveBeenCalledWith(
        defaultProps.orgId
      );
      expect(defaultProps.$services.OrganizationRoles.isOwnerOrAdmin).toHaveBeenCalledWith(testOrg);
      const errArg = defaultProps.onForbidden.mock.calls[0][0];
      expect(errArg).toBeInstanceOf(Error);
      expect(errArg.message).toBe('No permission');
    });
  });

  describe('fetching org data fails with 404', () => {
    const error404 = new Error('Test error');
    error404.status = 404;

    beforeEach(() => {
      defaultProps.$services.OrganizationMembershipRepository.getAllSpaces.mockReturnValue(
        Promise.reject(error404)
      );
    });

    it('should call `onForbidden`', async () => {
      await shallowRenderComponent(defaultProps);

      expect(defaultProps.onForbidden).toHaveBeenCalledWith(error404);
    });
  });

  describe('fetching org data fails with 403', () => {
    const error403 = new Error('Test error');
    error403.status = 403;

    beforeEach(() => {
      defaultProps.$services.OrganizationMembershipRepository.getAllSpaces.mockReturnValue(
        Promise.reject(error403)
      );
    });

    it('should call `onForbidden`', async () => {
      await shallowRenderComponent(defaultProps);

      expect(defaultProps.onForbidden).toHaveBeenCalledWith(error403);
    });
  });

  describe('fetching org data fails with different error code', () => {
    const error400 = new Error('Test error');
    error400.status = 400;

    beforeEach(() => {
      defaultProps.$services.OrganizationMembershipRepository.getAllSpaces.mockReturnValue(
        Promise.reject(error400)
      );
    });

    it('should trigger reload notification', async () => {
      await shallowRenderComponent(defaultProps);

      expect(defaultProps.onForbidden).not.toHaveBeenCalled();
      expect(defaultProps.$services.ReloadNotification.trigger).toHaveBeenCalled();
    });
  });
});

describe('WorkbenchActions', () => {
  it('should render', () => {
    const wrapper = shallow(<WorkbenchActions />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('isLoading', () => {
    it('should render a spinner', () => {
      const wrapper = shallow(<WorkbenchActions isLoading />);

      expect(wrapper.find(Spinner)).toHaveLength(1);

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
    });
  });

  describe('org is committed, flag is set and periods received', () => {
    it('should render the PeriodSelector', () => {
      const wrapper = shallow(
        <WorkbenchActions
          hasSpaces
          committed
          flagActive
          periods={[]}
          selectedPeriodIndex={0}
          setPeriodIndex={() => {}}
        />
      );

      expect(wrapper.find(PeriodSelector)).toHaveLength(1);
    });
  });

  describe('org is not committed', () => {
    it('should render nothing', () => {
      const wrapper = shallow(
        <WorkbenchActions
          hasSpaces
          flagActive
          periods={[]}
          selectedPeriodIndex={0}
          setPeriodIndex={() => {}}
        />
      );

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
      expect(wrapper.find(Spinner)).toHaveLength(0);
    });
  });

  describe('org has no spaces', () => {
    it('should render nothing', () => {
      const wrapper = shallow(
        <WorkbenchActions
          committed
          flagActive
          periods={[]}
          selectedPeriodIndex={0}
          setPeriodIndex={() => {}}
        />
      );

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
      expect(wrapper.find(Spinner)).toHaveLength(0);
    });
  });

  describe('feature flag not set', () => {
    it('should render nothing', () => {
      const wrapper = shallow(
        <WorkbenchActions
          hasSpaces
          committed
          periods={[]}
          selectedPeriodIndex={0}
          setPeriodIndex={() => {}}
        />
      );

      expect(wrapper.find(PeriodSelector)).toHaveLength(0);
      expect(wrapper.find(Spinner)).toHaveLength(0);
    });
  });
});

describe('WorkbenchContent', () => {
  let defaultProps = null;

  beforeEach(() => {
    defaultProps = {
      committed: true,
      flagActive: true,
      hasSpaces: true,
      selectedPeriodIndex: 0,
      spaceNames: { space1: 'Space1', space2: 'Space2' },
      isPoC: { space1: false, space2: true },
      periodicUsage: {},
      apiRequestIncludedLimit: 1000,
      assetBandwidthUsage: 100,
      assetBandwidthIncludedLimit: 50,
      assetBandwidthUOM: 'GB',
      isLoading: false,
      periods: [],
      resources: []
    };
  });

  it('should render', () => {
    const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('org is committed, flag is active and there are spaces', () => {
    it('should render the OrganizationUsagePage', () => {
      const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

      expect(wrapper.find(OrganizationUsagePage)).toHaveLength(1);

      expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(0);
      expect(wrapper.find(OrganizationResourceUsageList)).toHaveLength(0);
    });
  });

  describe('org has no spaces', () => {
    it('should render NoSpacePlaceholder', () => {
      const wrapper = shallow(
        <WorkbenchContent {...{ ...defaultProps, ...{ hasSpaces: false } }} />
      );

      expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(1);

      expect(wrapper.find(OrganizationUsagePage)).toHaveLength(0);
      expect(wrapper.find(OrganizationResourceUsageList)).toHaveLength(0);
    });
  });

  describe('org is not committed', () => {
    beforeEach(() => {
      defaultProps.committed = false;
    });

    it('should render OrganizationResourceUsageList', () => {
      const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

      expect(wrapper.find(OrganizationResourceUsageList)).toHaveLength(1);

      expect(wrapper.find(OrganizationUsagePage)).toHaveLength(0);
      expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(0);
    });

    describe('no resources given', () => {
      beforeEach(() => {
        defaultProps.resources = undefined;
      });

      it('should render nothing', () => {
        const wrapper = shallow(<WorkbenchContent {...defaultProps} />);

        expect(wrapper.find(OrganizationResourceUsageList)).toHaveLength(0);
        expect(wrapper.find(OrganizationUsagePage)).toHaveLength(0);
        expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(0);
      });
    });
  });

  describe('feature flag is not active', () => {
    it('should render OrganizationResourceUsageList', () => {
      const wrapper = shallow(
        <WorkbenchContent {...{ ...defaultProps, ...{ flagActive: false } }} />
      );

      expect(wrapper.find(OrganizationResourceUsageList)).toHaveLength(1);

      expect(wrapper.find(OrganizationUsagePage)).toHaveLength(0);
      expect(wrapper.find(NoSpacesPlaceholder)).toHaveLength(0);
    });
  });
});
