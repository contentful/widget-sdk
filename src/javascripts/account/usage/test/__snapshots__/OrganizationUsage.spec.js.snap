// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationUsage should render page without errors 1`] = `
<Workbench
  testId="organization.usage"
>
  <WorkbenchHeader
    actions={
      <WorkbenchActions
        committed={true}
        error={null}
        hasSpaces={true}
        isLoading={true}
        periods={
          Array [
            Object {
              "endDate": null,
              "startDate": "2018-12-23",
              "sys": Object {
                "id": "0",
                "type": "UsagePeriod",
              },
            },
            Object {
              "endDate": "2018-12-22",
              "startDate": "2018-11-22",
              "sys": Object {
                "id": "1",
                "type": "UsagePeriod",
              },
            },
          ]
        }
        setPeriodIndex={[Function]}
      />
    }
    icon={
      <Icon
        name="page-usage"
        scale="0.75"
      />
    }
    testId="cf-ui-workbench-header"
    title="Usage"
  />
  <WorkbenchContent
    testId="cf-ui-workbench-content"
    type="default"
  >
    <WorkbenchContent
      apiRequestIncludedLimit={1000000}
      committed={true}
      error={null}
      hasSpaces={true}
      isLoading={true}
      isPoC={
        Object {
          "space1": false,
          "space2": false,
        }
      }
      periods={
        Array [
          Object {
            "endDate": null,
            "startDate": "2018-12-23",
            "sys": Object {
              "id": "0",
              "type": "UsagePeriod",
            },
          },
          Object {
            "endDate": "2018-12-22",
            "startDate": "2018-11-22",
            "sys": Object {
              "id": "1",
              "type": "UsagePeriod",
            },
          },
        ]
      }
      spaceNames={
        Object {
          "test1": "Test1",
          "test2": "Test2",
        }
      }
    />
  </WorkbenchContent>
</Workbench>
`;

exports[`WorkbenchActions should render 1`] = `""`;

exports[`WorkbenchContent should render 1`] = `
<OrganizationUsagePage
  apiRequestIncludedLimit={1000}
  isLoading={false}
  isPoC={
    Object {
      "space1": false,
      "space2": true,
    }
  }
  periodicUsage={Object {}}
  spaceNames={
    Object {
      "space1": "Space1",
      "space2": "Space2",
    }
  }
/>
`;
